core/src/main/java/io/grpc/internal/Channelz.java

```diff
@@ -359,20 +360,79 @@ public ChannelStats build() {
     public final SocketAddress local;
     public final SocketAddress remote;
     public final Security security;
+    public final SocketOptions socketOptions;
 
     /** Creates an instance. */
     public SocketStats(
         TransportStats data,
         SocketAddress local,
         SocketAddress remote,
+        SocketOptions socketOptions,
         Security security) {
       this.data = data;
       this.local = local;
       this.remote = remote;
+      this.socketOptions = socketOptions;
       this.security = security;
     }
   }
 
+  public static final class SocketOptions {
+    public final Map<String, String> others;
+    // In netty, the value of a channel option may be null.
+    @Nullable public final Integer soTimeoutMillis;
+    @Nullable public final Integer lingerSeconds;
+
+    /** Creates an instance. */
+    public SocketOptions(
+        Integer timeoutMillis,
+        Integer lingerSeconds,
+        Map<String, String> others) {
+      this.soTimeoutMillis = timeoutMillis;
+      this.lingerSeconds = lingerSeconds;
+      this.others = Collections.unmodifiableMap(Preconditions.checkNotNull(others));
```

Comment:
This won't actually be immutable.  Adding things to the builder *after* building options will modify the SocketOptions.   
```
b = new Builder();
so = b.build();
b.addOption(a, b);
// so changed!
```